File: pages/LoginPage.ts
import { type Locator, type Page } from '@playwright/test';

// Page Object de la pantalla de inicio de sesión de saucedemo.com.
// Es el control remoto de esa pantalla: las pruebas usan sus métodos y no escriben localizadores.
export class LoginPage {
  // Pestaña del navegador con la que trabaja la página. La recibe el constructor.
  private readonly page: Page;

  // Campo donde se escribe el usuario. Lo buscamos por el texto gris que muestra adentro.
  private readonly campoUsuario: Locator;

  // Campo donde se escribe la clave, también por su texto gris.
  private readonly campoClave: Locator;

  // Botón "Login". Lo buscamos como lo ve una persona: un botón que dice Login.
  private readonly botonLogin: Locator;

  // Mensaje rojo que sale cuando el login falla. Es público para que la prueba pueda revisar su texto.
  readonly mensajeError: Locator;

  // Recibe la pestaña y prepara los localizadores. Preparar un localizador no busca nada todavía.
  constructor(page: Page) {
    this.page = page;
    this.campoUsuario = page.getByPlaceholder('Username');
    this.campoClave = page.getByPlaceholder('Password');
    this.botonLogin = page.getByRole('button', { name: 'Login' });
    this.mensajeError = page.getByTestId('error');
  }

  // Abre la pantalla de login. La dirección completa sale de baseURL en playwright.config.ts.
  async abrir() {
    await this.page.goto('/');
  }

  // Escribe el usuario y la clave y pulsa Login.
  async iniciarSesion(usuario: string, clave: string) {
    await this.campoUsuario.fill(usuario);
    await this.campoClave.fill(clave);
    await this.botonLogin.click();
  }
}

File: src/main/java/co/com/semillero/playwright/pages/LoginPage.java
package co.com.semillero.playwright.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.AriaRole;

// Page Object de la pantalla de inicio de sesión de saucedemo.com.
// Las pruebas no ven los localizadores: solo llaman abrir(), iniciarSesion() y mensajeDeError().
public class LoginPage extends PaginaBase {

    // Dirección de la pantalla de login.
    private static final String DIRECCION = "https://www.saucedemo.com/";

    // Campo donde se escribe el usuario. Lo buscamos por el texto gris que muestra adentro.
    private final Locator campoUsuario;

    // Campo donde se escribe la clave, también por su texto gris.
    private final Locator campoClave;

    // Botón "Login". Lo buscamos como lo ve una persona: un botón que dice Login.
    private final Locator botonLogin;

    // Mensaje rojo que sale cuando el login falla. Lo buscamos por su atributo data-test.
    private final Locator mensajeError;

    // Recibe la pestaña, se la pasa a PaginaBase y prepara los localizadores.
    // Preparar un localizador no busca nada todavía: Playwright lo busca cuando lo usas.
    public LoginPage(Page pagina) {
        super(pagina);
        campoUsuario = pagina.getByPlaceholder("Username");
        campoClave = pagina.getByPlaceholder("Password");
        botonLogin = pagina.getByRole(AriaRole.BUTTON, new Page.GetByRoleOptions().setName("Login"));
        mensajeError = pagina.locator("[data-test='error']");
    }

    // Abre la pantalla de login.
    public void abrir() {
        navegarA(DIRECCION);
    }

    // Escribe el usuario y la clave y pulsa Login.
    public void iniciarSesion(String usuario, String clave) {
        campoUsuario.fill(usuario);
        campoClave.fill(clave);
        botonLogin.click();
    }

    // Devuelve el texto del mensaje de error para que la prueba lo compare.
    public String mensajeDeError() {
        return mensajeError.textContent();
    }
}
